Postgres raster temporal API support

elpaso opened the pull request:
## Temporal capabilities integration for Postgres rasters

This PR also contains a few bug fixes and speed optimizations, I'm planning to add a some more test cases before the final merge.

![qgis-pg-raster-temporal](https://user-images.githubusercontent.com/142164/78421830-1f73b680-765b-11ea-8e65-6c2040ad24f2.gif)


![qgis-pgraster-temporal-default](https://user-images.githubusercontent.com/142164/78421839-37e3d100-765b-11ea-83e5-e7e3e519155e.png)

Funded by: **ARPA Piemonte**

elpaso committed:
Const correct provider temporal capabilities

elpaso committed:
PG raster: temporal support

elpaso committed:
Fix encoded URI

elpaso committed:
Fix datasource URI for WMS-T

elpaso committed:
Const correctness

elpaso committed:
Fix WMS-T temporal data source URI

elpaso commented:
@Samweli see my last commits, I cannot see any error here after my last commits, but you probably know how to reproduce the issues better than me.

Samweli commented:
> @Samweli see my last commits, I cannot see any error here after my last commits, but you probably know how to reproduce the issues better than me.

@elpaso I have tested, I run into different issues, one is layer datasource is not available after static temporal range update.
![Error_wms_temporal_update](https://user-images.githubusercontent.com/2663775/78190788-5a1ff800-747d-11ea-975a-bf056a8b0655.png)

nyalldawson commented on python/core/auto_generated/qgsdataprovider.sip.in:
```diff
@@ -116,7 +116,7 @@ Gets the data source specification.
 %End
 
-    virtual QgsDataProviderTemporalCapabilities *temporalCapabilities();
+    virtual QgsDataProviderTemporalCapabilities *temporalCapabilities() const;
```
There's discussion about this here: https://github.com/qgis/QGIS/pull/35466#discussion_r401269274

nyalldawson commented on src/providers/postgres/raster/qgspostgresrasterprovider.cpp:
```diff
@@ -1185,6 +1228,46 @@ bool QgsPostgresRasterProvider::init()
+  // Setup temporal properties for layer
+  if ( mUri.hasParam( QStringLiteral( "temporalField" ) ) )
+  {
+    Q_ASSERT_X( temporalCapabilities(), "QgsPostgresRasterProvider::QgsPostgresRasterProvider", "Data provider temporal capabilities object does not exist" );
```
I don't think this is explicitly required -- the temporalCapabilities object is always created by QgsRasterDataProvider

nyalldawson reviewed (commented):
Looking good to me! 3.14 will be when QGIS enters the 4th dimension :smile: 

One thing you may want to consider -- the wms provider exposes the option of what to do when an exact time range match isn't possible. E.g. picking the closest possible value, using the time from the start of the range, the end of the range, etc. I think those options could also be useful here.

elpaso commented:
@Samweli I cannot reproduce, can you please make sure you are testing exactly this branch? If you still have issues can you upload a project and the exact steps to reproduce the issue?

Samweli commented:
> @Samweli I cannot reproduce, can you please make sure you are testing exactly this branch? If you still have issues can you upload a project and the exact steps to reproduce the issue?

Yes I'm using this branch, I have attached below screenshot of steps to reproduce invalid projection and disabled reference time option issues for a layer which have reference time in capabilities.
I have used https://geo.weather.gc.ca/geomet?layer=GDPS.ETA_TT layer.

![wms_temporal_range_update_error](https://user-images.githubusercontent.com/2663775/78231943-393abf80-74dc-11ea-834e-714bc8c88b7c.gif)

elpaso committed:
Add dataComment to data provider base class

elpaso committed:
Add override after dataComment move

elpaso committed:
Added fields accessor to raster data provider

elpaso committed:
Synchonize provider data source URI

this is a bugfix (unreported?)

elpaso committed:
Added some consts

elpaso committed:
More PG raster temporal support

- expose to GUI
- allows to set a default datetime
- tests
- fix metadata retrieval for multiband filtered rasters

elpaso committed:
Fix temporal range slices (exclude end)

elpaso committed:
PG raster temporal, partly revert expose default as filter

elpaso committed:
Micro optimizations

elpaso committed:
PG raster: optimize slow query

elpaso committed:
PG raster temporal: cleanup after rebase conflicts

elpaso committed:
PG raster: need for speed for untiled rasters

elpaso committed:
Spellcheck ignore provider sql files

elpaso committed:
PG raster: fix raster size for untiled no metadata

Fix unreported issue with a wierd raster that has
uncomplete metadata:

- wrong size 1px rounding error
- nodata values were not cleared after dropping out from
  the fast track

elpaso committed:
Small typo in a comment

timlinux commented:
So exciting to see this @elpaso ! Do you have a sample dataset / table dump that we can play with?

elpaso commented:
@timlinux there isfew sample datasets in the testdata folder here: 

https://raw.githubusercontent.com/qgis/QGIS/0351cb203d58ca2492bfd99c693a96453bd223d0/tests/testdata/provider/postgresraster/raster_4326_time.sql

But I'm going to replace it soon with a smaller dataset for the CI tests.

elpaso committed:
PG raster: simplify tests data

Samweli commented:
> @elpaso I have tested, I run into different issues, one is layer datasource is not available after static temporal range update.
> ![Error_wms_temporal_update](https://user-images.githubusercontent.com/2663775/78190788-5a1ff800-747d-11ea-975a-bf056a8b0655.png)

Hi @elpaso, I'm still getting the above issue.
